spec/models/projects/customizable_spec.rb

```diff
@@ -27,92 +27,407 @@
     end
-    let(:custom_values) { [bool_custom_value, text_custom_value] }
 
-    subject { project.custom_value_attributes }
+    context 'with #custom_field_values= method' do
+      before do
+        project.custom_field_values = {
+          text_custom_field.id => 'foo',
+          bool_custom_field.id => true
+        }
+
+        project.save!
+      end
+
+      it_behaves_like 'implicitly enabled and saved custom values'
+    end
+
+    it 'does not re-enable fields without new value which have been disabled in the past (regression)' do
+      project.update!(custom_field_values: {
+                        text_custom_field.id => 'foo',
+                        bool_custom_field.id => true
+                      })
+
+      expect(project.reload.project_custom_fields)
+        .to contain_exactly(text_custom_field, bool_custom_field)
+
+      project.project_custom_field_project_mappings.find_by(custom_field_id: text_custom_field.id).destroy
+
+      expect(project.reload.project_custom_fields)
+        .to contain_exactly(bool_custom_field)
+
+      project.update!(custom_field_values: {
+                        bool_custom_field.id => true
+                      })
+
+      expect(project.reload.project_custom_fields)
+        .to contain_exactly(bool_custom_field)
+    end
+
+    context 'with correct handling of custom fields with default values' do
+      let!(:text_custom_field_with_default) do
+        create(:text_project_custom_field,
+               default_value: 'default',
+               project_custom_field_section: section)
+      end
+
+      it 'does not activate custom fields with default values if not explicitly set to a value' do
+        project.update!(custom_field_values: {
+                          text_custom_field.id => 'bar',
+                          bool_custom_field.id => false
+                        })
+
+        # text_custom_field_with_default is not provided, thus it should not be enabled (in contrast to creation)
+        expect(project.project_custom_field_project_mappings.pluck(:custom_field_id))
+          .to contain_exactly(text_custom_field.id, bool_custom_field.id)
+      end
+
+      it 'does activate custom fields with default values if explicitly set to a value' do
+        project.update!(custom_field_values: {
+                          text_custom_field.id => 'bar',
+                          bool_custom_field.id => false,
+                          text_custom_field_with_default.id => 'overwritten default'
+                        })
+
+        # text_custom_field_with_default is not provided, thus it should not be enabled (in contrast to creation)
+        expect(project.project_custom_field_project_mappings.pluck(:custom_field_id))
+          .to contain_exactly(text_custom_field.id, bool_custom_field.id, text_custom_field_with_default.id)
+      end
+    end
+
+    it 'does re-enable fields with new value which have been disabled in the past' do
```

Comment:
Is this the usecase to ensure that updates coming from the api will accept the disabled custom fields too? I'm wondering if we should raise a validation error instead?

Reply:
I figured that's the way we enable the implicit control over the activation state via the API without further API adjustments.